File: Array/caiweixiong/java/Homework_011.java
import java.util.Queue;

public class Homework_011 {

    // 用队列实现栈，LeetCode第225题
    void homework_011_225() {
        MyStack obj = new MyStack();
        obj.push(4);
        int param_2 = obj.pop();
        System.out.println(param_2);
        int param_3 = obj.top();
        System.out.println(param_3);
        boolean param_4 = obj.empty();
        System.out.println(param_4);
    }

    class MyStack {

        private Queue mStack;

        /**
         * Initialize your data structure here.
         */
        public MyStack() {
            mStack = new Queue();
            mStack.front = null;
            mStack.rear = null;
        }

        /**
         * Push element x onto stack.
         */
        public void push(int x) {
            Node node = new Node();
            node.data = x;
            node.next = null;
            if (mStack.front == null && mStack.rear == null)
                mStack.front = mStack.rear = node;
            else {
                node.next = mStack.front;
                mStack.front = node;
            }
        }

        /**
         * Removes the element on top of the stack and returns that element.
         */
        public int pop() {
            int ele = 0;
            if (mStack.front == null)
                return ele;
            ele = mStack.front.data;
            if (mStack.front == mStack.rear)
                mStack.front = mStack.rear = null;
            else
                mStack.front = mStack.front.next;
            return ele;
        }

        /**
         * Get the top element.
         */
        public int top() {
            if (mStack.front == null)
                return 0;
            else
                return mStack.front.data;
        }

        /**
         * Returns whether the stack is empty.
         */
        public boolean empty() {
            return mStack.front == null;
        }
    }

    class Node {
        int data;
        Node next;
    }

    class Queue {
        Node front;
        Node rear;
    }
}

File: Array/caiweixiong/c/homework_011.c
//
// Created by Vashon on 2020/6/15.
//

#include "list_node.h"
#include <stdlib.h>
#include <stdio.h>

// 用队列实现栈，LeetCode第225题
void homework_011_225(void) {
    MyStack *myStackCreate();
    void myStackPush(MyStack *, int);
    int myStackPop(MyStack *obj);
    int myStackTop(MyStack *obj);
    int myStackEmpty(MyStack *obj);
    void myStackFree(MyStack *obj);

    MyStack *stack = myStackCreate();
    myStackPush(stack, 3);
    int param_2 = myStackPop(stack);
    int param_3 = myStackTop(stack);
    int param_4 = myStackEmpty(stack);
    printf("param_2:%d, param_3:%d, param_4:%d\n", param_2, param_3, param_4);
    myStackFree(stack);
}

// 初始化栈
MyStack *myStackCreate() {
    MyStack *stack = malloc(sizeof(MyStack));
    stack->front = NULL;
    stack->rear = NULL;
    return stack;
}

// 压栈
void myStackPush(MyStack *obj, int x) {
    // 栈后进先出的特性，所以，头插法建立队列，就能最大程度的优化栈
    if (!obj)
        return;
    // 创建新节点
    Node *node = malloc(sizeof(Node));
    node->next = NULL;
    node->val = x;
    if ((!obj->rear) && (!obj->front))    // 首次创建节点
        obj->rear = obj->front = node;
    else {  // 再次创建，头插法建队
        node->next = obj->front;
        obj->front = node;
    }
}

// 弹栈，并返回弹出元素
int myStackPop(MyStack *obj) {
    int ele = 0;
    if (!obj || ((!obj->rear) && (!obj->front)))    // 栈空时，返回0
        return ele;
    ele = obj->front->val;
    Node *node = obj->front;        // 取出队首节点
    if (obj->front == obj->rear) {  // 只有一个元素
        obj->front = NULL;
        obj->rear = NULL;
    } else {
        obj->front = node->next;
    }
    free(node);
    return ele;
}

// 获取栈顶元素
int myStackTop(MyStack *obj) {
    if (!obj || ((!obj->rear) && (!obj->front)))
        return 0;
    else
        return obj->front->val;
}

// 栈是否为空
int myStackEmpty(MyStack *obj) {
    if (!obj || ((!obj->rear) && (!obj->front)))
        return 1;
    else
        return 0;
}

// 释放栈
void myStackFree(MyStack *obj) {
    if (!obj)
        return;
    Node *node = obj->front;
    while (node) {
        obj->front = node->next;
        free(node);
        node = obj->front;
    }
    free(obj);
}